### PR TITLE
Global config to filter messages out

### PR DESCRIPTION
**Pull request by santiagodoldan**

### CODE REVIEW

**Comment by santiagodoldan** [comment]
https://github.com/microsoftly/BotTester/issues/15

**Comment by santiagodoldan** [comment]
@microsoftly 

About how `MessageComparator.ts` compares messages, I decided to not use recursive functions because  the code ends a bit messy and difficult to understand, but if you don't like that, I can change it and use recursive functions to detect differences between 2 message objects.

This is just a WIP that we are going to use in our project because we are having lot of issues with messages of type `endOfConversation` and `__save__`, we've created a custom Prompt that uses `session.save()` and it sends a silent message of type `__save__`, without a filter, we need to add an expect for that message in almost all the dialogs that our bot has.

Also, @bilby91 please review it too 🌮

**Comment by santiagodoldan** [comment]
I've just refactored the `isEqual` implementation of the new `MessageComparator` class, talked with @bilby91 about it and the solution is not as complex as I expected.

File: test/BotTester.spec.ts [hunk 1]
```diff
@@ -321,6 +321,20 @@ describe('BotTester', () => {
     });
 //```
 
+//```javascript
+    it('can filter messages out', () => {
```
Comment (microsoftly) [inline test/BotTester.spec.ts]: copy this section of code to the README.md. Theres no auto-gen script at the moment

File: src/MessageComparator.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+export class MessageComparator {
+  private message: IMessage;
+
+  constructor(message: IMessage) {
+    this.message = message;
+  }
+
+  public isEqual(msg: IMessage): boolean {
+    return (function deepInclude(message: any, otherMessage: any): boolean {
```
Comment (microsoftly) [inline src/MessageComparator.ts]: I'd recommend making a private function that you call here. This is not immediately clear what is happening

**Comment by microsoftly** [comment]
@santiagodoldan 

Theres a few points to address here: 

* tests in BotTester.spec.ts that demonstrate functionality need to be copied to the README.md manually. Keeping the //```javascript comment is there for easy copying later for large overhauls
* consistent style: 4 space indenting, semicolons to end lines, etc ... (I may have missed some cases)
* the ```MessageComparator``` and ```IConfig.filteredMessages``` seems like a lot of structure without much flexibility. Instead, I'd recommend modifying the config with something like this:
```typescript
interface IConfig {
// ...
    // returns true for messages that the BotTester should not ignore. False for messages it should ignore 
    messageFilter: (message: IMessage) => boolean
}
```

then use that later to filter incoming messages by using the function like so [here](https://github.com/microsoftly/BotTester/blob/master/src/MessageService.ts#L108)

```typescript
const filteredMessages = e.filter(messageFilter)

this.botToUserMessageChecker(filteredMessages);
```

doing this makes the filter more flexible and fit the needs of the end user, including cases we might not think of. It also makes it easier to test and maintain.

After this feature is merged, we can make a separate PR with an additional config option of ```shouldSendSilentSave: boolean```  and any additional cases that may be generically useful (e.g. ```shouldIgnoreEndOfConversationEvent```,  ```shouldIgnoreTypingEvent```, etc ...)

**Comment by santiagodoldan** [comment]
@microsoftly

I like the `messageFilter` approach, but we would need to find a way to make it global (if not, each BotTester instance will need to receive the function), the nice thing of having a `filteredMessages` array of objects, is that it can be stored in the `bot-tester.yml` config file you added, here you have an example of how it looks that file in our project now:

```
{
  "timeout": -1,
  "defaultAddress": {
      "channelId": "console",
      "user": { "id": "user1", "name": "user1" },
      "bot": { "id": "bot", "name": "Bot" },
      "conversation": { "id": "user1Conversation" }
  },
  "filteredMessages": [
    { "type": "endOfConversation" },
    { "type": "__save__" }
  ]
}
```

Saying that, I'm thinking that maybe adding a new function to BotTester that defines the `messageFilter` globally could be a good idea, so it can be defined in the `spec_helper.js` only once.

```
BotTester
  .filter((message) => {
    return ['endOfConversation', '__save__'].includes?(message.type);
  });
```

**Comment by microsoftly** [comment]
For your cases, I think what I've described works fine. We once the filter is implemented, we can add in some standard options into the config e.g. a config file with:
``` typescript
{
  ignoreInternalSaveMessage: true,
  ignoreEndOfConversationEvent: true,
  ignoreTypingEvent: true
}
```

these will activate prebuilt filters

**Comment by santiagodoldan** [comment]
@microsoftly I've just removed all the `MessageComparator` stuff, instead of all that complex logic, 
 now there's a simple array of strings that contains message types to be filtered out. (`filteredMessageTypes: ["__save__", "endOfConversation"]`).

The library keeps simple and I can think a bit more about the filter function.

I'm gonna be out for a few days, will try to think a bit more about that filter function so I can create a new PR and add that too.

**Comment by microsoftly** [comment]
I've realized that the fact that you're seeing the internal ``` __save__``` message is a bug. I'm taking this over and will fix the bug as well

**Comment by santiagodoldan** [comment]
__save__ is not the only one causing issues in our tests, what are you planning to do?

**Comment by microsoftly** [comment]
I've implemented the solution. You will just need to add your own custom message filters as needed.

I decided not to make a global filter. One possible reason is that you want to ignore the ```__save__``` messages now, but ignoring them will break the checkSession functionality. Down the line you may want to check session, and need to update everything.

Additionally, the options are modifiers to the default functionality. The default functionality is sending the ```___save__``` (this is default functionality because it is at the core of checking session state) message along with all other types of event messages. For that reason, it seemed more reasonable to keep them as such.

**Comment by santiagodoldan** [comment]
This PR does not filter any messages by default :/ by default it's an empty array.

If I need a message in a specific test because I filtered that message type using the bot-tester.yml config file, then I can change that using the constructor, because it overrides the global configurations. The same as the defaultAddress.

Will try your changes on Monday, I'm on vacation now, I'll let you know then.

**Comment by microsoftly** [comment]
you can filter some standard messages by default (e.g. https://github.com/microsoftly/BotTester/blob/master/src/config.ts#L28) If you need more messages that can be standardly ignored, they can be added to the config later.

**Comment by bilby91** [comment]
@santiagodoldan I think that configuration should work for our use case at the moment so we are covered. 

FIY @microsoftly  their is a typo in the config `ignoreInternalSveMessage` => `ignoreInternalSaveMessage`

**Pull request closed without merging**